aten/src/ATen/core/Tensor.cpp

```diff
@@ -92,9 +93,183 @@ const Tensor& Tensor::requires_grad_(bool _requires_grad) const {
+      return "swapaxes";
+    case ViewMeta::Type::swapdims:
+      return "swapdims";
+    case ViewMeta::Type::unfold:
+      return "unfold";
+    case ViewMeta::Type::alias:
+      return "alias";
+    case ViewMeta::Type::view_as_real:
+      return "view_as_real";
+    case ViewMeta::Type::view_as_complex:
+      return "view_as_complex";
+    case ViewMeta::Type::_conj:
+      return "_conj";
+    case ViewMeta::Type::_neg_view:
+      return "_neg_view";
+    case ViewMeta::Type::chunk:
+      return "chunk";
+    case ViewMeta::Type::diagonal:
+      return "diagonal";
+    case ViewMeta::Type::diagonal_Dimname:
+      return "diagonal_Dimname";
+    case ViewMeta::Type::expand_as:
+      return "expand_as";
+    case ViewMeta::Type::narrow:
+      return "narrow";
+    case ViewMeta::Type::narrow_Tensor:
+      return "narrow_Tensor";
+    case ViewMeta::Type::movedim_intlist:
+      return "movedim_intlist";
+    case ViewMeta::Type::movedim_int:
+      return "movedim_int";
+    case ViewMeta::Type::reshape_as:
+      return "reshape_as";
+    case ViewMeta::Type::detach:
+      return "detach";
+    case ViewMeta::Type::transpose_int:
+      return "transpose_int";
+    case ViewMeta::Type::transpose_Dimname:
+      return "transpose_Dimname";
+    case ViewMeta::Type::unsqueeze:
+      return "unsqueeze";
+    case ViewMeta::Type::_values:
+      return "_values";
+    case ViewMeta::Type::indices:
+      return "indices";
+    case ViewMeta::Type::view:
+      return "view";
+    case ViewMeta::Type::view_dtype:
+      return "view_dtype";
+      default:
+        return "UNKNOWN_VIEW_TYPE";
+  }
+}
+
+std::ostream& operator<<(std::ostream& str, ViewMeta::Type rhs) {
+  return str << ViewMetaTypeToString(rhs);
+}
+
+Alias::Alias(const at::Tensor& base) : base_(base.clone()) {
+  auto impl = dynamic_cast<at::FunctionalTensorImplBase*>(base_.unsafeGetTensorImpl());
+  TORCH_INTERNAL_ASSERT_DEBUG_ONLY(impl != nullptr);
+  // Note [Marking Alias Tensors]
+  // In the functionalization pass, calling a view operation on a tensor forks it off an Alias object
+  // which is shared between the original tensor and the view.
+  // The Alias object contains a base tensor, which is used to sync operations across the view tensors.
+  // Throughout the Alias' life cylc,e we need to perform various views and reshapes on the base alias tensor.
+  // We don't want that process to recursively fork off another Alias object though, so we explicitly mark
+  // the base tensor as an alias.
+  // Another option would have been to implement separate {view}_copy() kernels for every corresponding view op,
+  // which have fallthrough kernels registered in the functionalization pass.
+  impl->mark_as_alias();
+}
+
+const at::Tensor& Alias::base() const {
+  return base_;
+}
+
+// metas is taken by value on purpose - we want to copy the vector.
+void Alias::add_update(const at::Tensor& updated_val, std::vector<ViewMeta> metas) {
```

Comment:
There needs to be some documentation about why delaying updates to alias is profitable; you could imagine a correct implementation that always updates `base_` every time some calls `add_update`.

If I recall correctly, XLA does this because in some cases it turns out that it is not necessary to update the base. Imagine the following situation:

```
x = y[1]
x.add_(2)
# ... y is never used again ...
```

In this case, being lazy about the base update is profitable because there is no need to update y's metadata.

Delaying the updates is not an unmitigated good, because if you continuously keep applying inplace operations without triggering a sync you will leak memory.

(maybe @ailzhang remembers precisely)